plugin/nugus_plugin.cpp

```diff
@@ -0,0 +1,398 @@
+        // P-controller to the joints for positions.
+        double gain = target.gain()
+                      * (target.id() == JointID::R_ANKLE_ROLL || target.id() == JointID::L_ANKLE_ROLL
+                             ? joint_ankle_roll_pid_factor
+                             : joint_pid_factor);
+
+        // Only update the controller if the gain changed
+        model->GetJointController()->SetPositionPID(joints[target.id()]->GetScopedName(), common::PID(gain));
+
+        auto duration = time_point_cast(target.time()) - std::chrono::steady_clock::now();
+
+        // Calculate the joint's velocity
+        double current_position = model->GetJointController()->GetPositions()[joints[target.id()]->GetScopedName()];
+        double position_difference =
+            M_PI - std::fabs(std::fmod(std::fabs(current_position - target.position()), 2.0 * M_PI) - M_PI);
+        double velocity =
+            position_difference
+            / (static_cast<double>(duration.count()) / static_cast<double>(std::chrono::steady_clock::period::den));
+
+        // Set the joint's target position.
+        model->GetJointController()->SetPositionTarget(joints[target.id()]->GetScopedName(), target.position());
+
+        if (velocity > 0.0) {
+            joints[target.id()]->SetVelocityLimit(0, velocity);
+        }
+        else {
+            joints[target.id()]->SetVelocityLimit(0, joint_velocity_limit);
+        }
+
+        joints[target.id()]->SetStiffnessDamping(0, joint_stiffness, joint_damping, joints[target.id()]->Position());
+    }
+
+    void update_robot() {
+
+        // Run all our commands in the command queue and then clear it
+        {
+            std::lock_guard<std::mutex> lock(command_mutex);
+            for (const auto& c : command_queue) {
+                apply_joint_command(c);
+            }
+
+            // Must be called every time update step to apply forces
+            if (!command_queue.empty()) {
+                model->GetJointController()->Update();
+                command_queue.resize(0);
+            }
+        }
+
+        // Get the current simulation time
+        double now = model->GetWorld()->SimTime().Double();
+
+        // If our last update was in the future then we reset our time so reset our last update
+        last_update = last_update > now ? now - update_rate : last_update;
+        if (now - last_update >= update_rate) {
+            last_update += update_rate;
+
+            auto msg = std::make_unique<message::platform::gazebo::RawSensors>();
+            msg->set_model(model->GetName());
+            auto sensors = msg->mutable_sensors();
+            for (uint32_t i = 0; i < 20; ++i) {
+                auto s = sensors->add_servos();
+                s->set_presentposition(joints[servo_id_to_joint[i]]->Position() + joint_offsets[servo_id_to_joint[i]]);
+                s->set_presentspeed(joints[servo_id_to_joint[i]]->GetVelocity(0));
+            }
```

Comment:
Any chance of getting other properties of the servo? Ideally, we need as many of the following as we can to ensure that no algorithms will do strange stuff.
```c++
         uint32 errorFlags      = 1;
         bool   torqueEnabled   = 2;
         float  pGain           = 3;
         float  iGain           = 4;
         float  dGain           = 5;
         float  goalPosition    = 6;
         float  movingSpeed     = 7;
         float  torque          = 8;
         float  presentPosition = 9;
         float  presentSpeed    = 10;
         float  load            = 11;
         float  voltage         = 12;
         float  temperature     = 13;
```